### PR TITLE
Make the text selection code more defensive

### PR DESCRIPTION
**Pull request by xster**
https://github.com/flutter/flutter/issues/12781

Make sure at no moment is the selected text range incompatible with the current text

### CODE REVIEW

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 1]
```diff
@@ -76,18 +76,30 @@ @interface FlutterTextRange : UITextRange<NSCopying>
...
 @end
 
 @implementation FlutterTextRange
 
 + (instancetype)rangeWithNSRange:(NSRange)range {
-  return [[[FlutterTextRange alloc] initWithNSRange:range] autorelease];
+  return [[[FlutterTextRange alloc] initWithNSRange:range withText:nil] autorelease];
+}
+
++ (instancetype)rangeWithNSRange:(NSRange)range withText:(NSString*)text {
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Given that the text itself isn't part for the state, `forText:` seems like the better name.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 2]
```diff
@@ -198,7 +210,7 @@ - (void)setTextInputState:(NSDictionary*)state {
   if (selectedRange.location != oldSelectedRange.location ||
       selectedRange.length != oldSelectedRange.length) {
     [self.inputDelegate selectionWillChange:self];
-    [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange]
+    [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange withText:self.text]
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Aren't we doing the newly added checks alreadyt on lines 206-209 above?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 3]
```diff
@@ -268,8 +280,11 @@ - (void)replaceRange:(UITextRange*)range withText:(NSString*)text {
     selectedRange.length -= intersectionRange.length;
   }
 
-  [self.text replaceCharactersInRange:replaceRange withString:text];
-  [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange]
+  [self.text
+      replaceCharactersInRange:[FlutterTextRange rangeWithNSRange:replaceRange withText:self.text]
+                                   .range
+                    withString:text];
+  [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange withText:self.text]
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Looks like we pass `nil` for all uses but this one and `setTextInputState`, and we're already doing these checks in `setTextInputState`. Maybe just add the min/max checks here and drop the changes to `FlutterTextRange`?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 4]
```diff
@@ -309,7 +324,7 @@ - (void)setMarkedText:(NSString*)markedText selectedRange:(NSRange)markedSelecte
 
   NSUInteger selectionLocation = markedSelectedRange.location + markedTextRange.location;
   selectedRange = NSMakeRange(selectionLocation, markedSelectedRange.length);
-  [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange]
+  [self setSelectedTextRange:[FlutterTextRange rangeWithNSRange:selectedRange withText:self.text]
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Isn't `setSelectedTextRange` already doing this check (twice now, maybe)?

Comment (xster) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: It's `setTextInputState` vs `setSelectedTextRange`, but I'll just extra the filtering into a static.

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm (hunk 1 above)
Comment (xster) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Deleted

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm (hunk 2 above)
Comment (xster) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Moved it out to a method

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm (hunk 3 above)
Comment (xster) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: Good idea, moved it out to a method

**Comment by chinmaygarde** [comment]
It seems like @xster has accounted for the all PR comments. @cbracken, is this PR good to go?

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm [hunk 5]
```diff
@@ -214,6 +214,12 @@ - (void)setTextInputState:(NSDictionary*)state {
   }
 }
 
+- (NSRange)validateSelection:(NSRange)range forText:(NSString*)text {
```
Comment (cbracken) [inline shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: nit: What about `clampSelection` (TBH I'm just hoping it gets some of those linewraps to squeak just under the column limit) ;)

**Comment by cbracken** [comment]
lgtm

File: shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm (hunk 5 above)
Comment (xster) [reply shell/platform/darwin/ios/framework/Source/FlutterTextInputPlugin.mm]: sure

**Pull request merged**